### PR TITLE
Improve function length property calculations

### PR DESCRIPTION
**Pull request by mmatyas**
Starting from ECMAScript 2015, the `length` property of functions
should be the number of parameters without a default value before
the first of such fields or a rest parameter.

See: https://www.ecma-international.org/ecma-262/11.0/index.html#sec-function-definitions-static-semantics-expectedargumentcount

JerryScript-DCO-1.0-Signed-off-by: Mátyás Mustoha mmatyas@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 1]
```diff
@@ -1491,16 +1491,38 @@ ecma_op_function_try_to_lazy_instantiate_property (ecma_object_t *object_p, /**<
...
+        literal_p -= args_p->register_end;
+        literal_p += args_p->literal_end;
+
         len = args_p->argument_end;
+        len_cbc_p = (uint8_t *) literal_p;
+      }
+
+      /* Function length is the number of non-initialized parameters, in which case
+       * there is an additional instruction at the beginning of the bytecode stream */
+      JERRY_ASSERT (len_cbc_p);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-function-object.c:1522]: JERRY_ASSERT (len_cbc_p != NULL);

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 2]
```diff
@@ -1491,16 +1491,38 @@ ecma_op_function_try_to_lazy_instantiate_property (ecma_object_t *object_p, /**<
...
+      if (len_cbc_p[0] == 0 && len_cbc_p[1] == CBC_EXT_SET_FUNCTION_LENGTH)
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-function-object.c:1523]: `len_cbc_p[0] == CBC_EXT_OPCODE`

File: jerry-core/parser/js/js-parser.c [hunk 3]
```diff
@@ -1806,6 +1806,15 @@ parser_parse_function_arguments (parser_context_t *context_p, /**< context */
 
 #if ENABLED (JERRY_ESNEXT)
   bool has_complex_argument = (context_p->next_scanner_info_p->u8_arg & SCANNER_FUNCTION_HAS_COMPLEX_ARGUMENT) != 0;
+
+  bool has_default_arguments = (context_p->next_scanner_info_p->u8_arg & SCANNER_FUNCTION_HAS_DEFAULT_ARGUMENTS) != 0;
+  uint8_t *cbc_fn_length_p = NULL;
+  if (has_default_arguments)
+  {
+    parser_emit_cbc_ext_call (context_p, CBC_EXT_SET_FUNCTION_LENGTH, 0xFF);
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser.c:1814]: I'd prefer `0` instead of `0xff`.

File: jerry-core/parser/js/js-parser.c [hunk 4]
```diff
@@ -1938,6 +1947,12 @@ parser_parse_function_arguments (parser_context_t *context_p, /**< context */
 
       context_p->status_flags |= PARSER_FUNCTION_HAS_COMPLEX_ARGUMENT;
 
+      if (cbc_fn_length_p)
+      {
+        *cbc_fn_length_p = (uint8_t) context_p->argument_count;
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser.c:1952]: Please assert here that `*cbc_fn_length_p` is `0`;

File: jerry-core/vm/vm.c [hunk 5]
```diff
@@ -2009,6 +2009,11 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
 
           goto free_left_value;
         }
+        case VM_OC_SET_FUNCTION_LENGTH:
+        {
+          byte_code_p++;
```
Comment (rerobika) [inline jerry-core/vm/vm.c:2014]: Please add a note about that during the execution this instruction is only a NOP.

File: jerry-core/vm/vm.c [hunk 6]
```diff
@@ -2009,6 +2009,11 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
 
           goto free_left_value;
         }
+        case VM_OC_SET_FUNCTION_LENGTH:
```
Comment (rerobika) [inline jerry-core/vm/vm.c:2012]: How about `VM_OC_EXPLICIT_FUNCTION_LENGTH`? (same for CBC instruction as well.)

File: tests/jerry/es.next/function-decl.js [hunk 7]
```diff
@@ -32,10 +32,10 @@ function f1(a,) {}
 assert(f1.length === 1)
 
 function f2(a = 1,) {}
-assert(f2.length === 1)
+assert(f2.length === 0)
 
 function f3(a = 1, b = 1 + 1, c,) {}
-assert(f3.length === 3)
+assert(f3.length === 0)
```
Comment (rerobika) [inline tests/jerry/es.next/function-decl.js:38]:
```js
var f5 = async(a = 1,) => {}
assert(f5.length === 1) // This should be zero as well
```

**Comment by mmatyas** [comment]
Ok, updated, and also added support for async and arrow functions.

**Comment by rerobika** [comment]
Closing after #4290 .

**Pull request closed without merging**